### ISSUE TITLE
Excessive memory usage using endpoint.MapControllerRoute - never ending story (continuation of #30092)

### ISSUE THREAD

**Issue by AlexanderKot** [dotnet/aspnetcore#38468]
### Describe the bug

Excessive memory usage on Asp.net Core application startup practically the same as described in #30092.

Our team in process of migration big enterprise level app from ASP.Net MVC5 (472) to .Net Core.
We have 414 controllers currently and approximately 55K actions (many generic controllers instantiated by IoC and so on). 
We first completed migration to .Core 3.1.21, and having it stabilized try migrate to  5.0.12.
There we have discovered behavior described in #30092 and continue migration to 6.0.

There are memory usage when app is fully initialized and showing login page
(functionality of Controllers\actions are same and routing works as predicted).

Approximate results :
| Framework | RAM | Top heap consumption  |
| --- | --- | --- |
|4.7.2   |~900m | In top position of memory dump we can see objects related to our code base. |
|3.1.21 |~2.7G | In top positions mainly MVC routing but our objects can be discovered. |
|5.0.12 |~**6.6G** | In top routing. |
|6.0 |~**7.4->5.7G** | In top routing. |

In 6.0 seems as improvement result related to mentioned issue after some idle time some memory released. 

At the same time we are exploring how to implement multi-tenancy where tenant code is first segment of url.
In my experiments this feature was commented, when not - routing info seems multiplied on tenants count
so for 2 tenants i have seen >10G after initialization in 5.0.12.  

So now we are actively exploring dynamic routing, 
\<sarcasm> as standard mechanism of routing so perfectly improved that cannot be used, except as in demo apps, where it demonstrate good performance. \</sarcasm>  

I understood all philosophy about RAM\CPU tradeoff and so on, but as for me such behavior is **huge regression** comparing to 3.1.21 (cannot say it about 4.72->3.1.21, as have not measured exact influence of routing in RAM consumption there, and migrating from 4.72 to it shows human visible improvement in app reaction).

Profiling our App on .Net 472 many times I have never seen any significant influence from routing mechanism, so cannot understood why it was needed to do this not optional "optimization" in routing with such RAM cost.

---------------
**.Net 472**

![47a](https://user-images.githubusercontent.com/1678749/142177930-86da4ab7-3ded-45b4-b553-6583e7ef8d3f.jpg)
![47b](https://user-images.githubusercontent.com/1678749/142178141-57892c4c-af74-4290-aa73-eacc2e6cea8a.jpg)
![47c](https://user-images.githubusercontent.com/1678749/142178244-83c3cc9d-0cbe-40b5-af6c-a8f0a245e6d3.jpg)

---------------
**.Net Core 3.1.21**

![3121a](https://user-images.githubusercontent.com/1678749/142178515-38df1b0a-8252-4e3a-aa78-da5a042ad458.jpg)
![3121c](https://user-images.githubusercontent.com/1678749/142178525-cd77a78a-9de3-49d5-ac27-ebb9b62f4e3b.jpg)

---------------
**.Net Core 5.0.12**

![5012a](https://user-images.githubusercontent.com/1678749/142178953-39fcdfe8-2c86-4296-ad6f-5f7eeab66523.jpg)
![5012c](https://user-images.githubusercontent.com/1678749/142178957-5ce7b689-a495-48a8-9489-992073fbd8b3.jpg)

---------------
**.Net Core 6.0**

![60a1](https://user-images.githubusercontent.com/1678749/142179117-3e9fa99a-f0e4-4feb-abdb-8168b8e85844.jpg)
![60c1](https://user-images.githubusercontent.com/1678749/142179124-38b66621-4f54-46d5-b2ac-e6c8f6586f98.jpg)

**Comment by javiercn** [comment]
@AlexanderKot thanks for contacting us.

We did some work in 6.0 to address this problem. We have some documentation [here](https://docs.microsoft.com/en-us/aspnet/core/fundamentals/routing?view=aspnetcore-6.0#guidance-for-large-route-tables) that describes simple changes you can do to your app to reduce the memory usage drastically.

**Comment by asamoylik** [comment]
We have constraints on parameters, but we can't move them to later segments of the path. Such memory usage appears when we have path with more than 4 segments (module/area/controller/action/id?). If such pattern is removed and MapDynamicControllerRoute is used, then memory usage falls down to ~2G.

**Comment by javiercn** [comment]
@asamoylik thanks for the additional details.

Could you share a sanitized version of your routes+controllers/pages so that we are able to provide some guidance wrt to next steps?

**Comment by asamoylik** [comment]
@javiercn here is the code of adding route patterns:
```
context.MapAreaControllerRoute("area1_Virtual", "area1", "{{module}}/area1/{{controller}}/{{action}}/{{id?}}", new { action = "index" }, new { module = new ModuleConstraint() });
context.MapAreaControllerRoute("area1_Default", "area1", "area1/{controller}/{action}/{id?}", new { action = "index", module = "area1m" });

context.MapAreaControllerRoute("area2_Virtual", "area2", "{{module}}/area2/{{controller}}/{{action}}/{{id?}}", new { action = "index" }, new { module = new ModuleConstraint() });
context.MapAreaControllerRoute("area2_Default", "area2", "area2/{controller}/{action}/{id?}", new { action = "index", module = "area2m" });

context.MapAreaControllerRoute("area3_Virtual", "area3", "{{module}}/area3/{{controller}}/{{action}}/{{id?}}", new { action = "index" }, new { module = new ModuleConstraint() });
context.MapAreaControllerRoute("area3_Default", "area3", "area3/{controller}/{action}/{id?}", new { action = "index", module = "area3m" });

context.MapAreaControllerRoute("area4_Virtual", "area4", "{{module}}/area4/{{controller}}/{{action}}/{{id?}}", new { action = "index" }, new { module = new ModuleConstraint() });
context.MapAreaControllerRoute("area4_Default", "area4", "area4/{controller}/{action}/{id?}", new { action = "index", module = "area4m" });

context.MapAreaControllerRoute("area5_Virtual", "area5", "{{module}}/area5/{{controller}}/{{action}}/{{id?}}", new { action = "index" }, new { module = new ModuleConstraint() });
context.MapAreaControllerRoute("area5_Default", "area5", "area5/{controller}/{action}/{id?}", new { action = "index", module = "area5m" });

context.MapAreaControllerRoute("area6_Virtual", "area6", "{{module}}/area6/{{controller}}/{{action}}/{{id?}}", new { action = "index" }, new { module = new ModuleConstraint() });
context.MapAreaControllerRoute("area6_Default", "area6", "area6/{controller}/{action}/{id?}", new { action = "index", module = "area6m" });

routes.MapControllerRoute("UnsupportedBrowser", "UnsupportedBrowser", new { controller = "UserAccount", action = "UnsupportedBrowser" });
routes.MapControllerRoute("DBVerIsObsolete", "UserAccount/LogOnToAdminPanel", new { controller = "UserAccount", action = "LogOnToAdminPanel" });
routes.MapControllerRoute("Default", "{controller}/{action}/{id?}",	new { controller = "UserAccount", action = "Logon" });
```
Paths with modules are used to organize virtual modules for user. Actions from areas can be used in different modules.

**Comment by AlexanderKot** [comment]
Hello
Is @asamoylik info enough or something else needed?

**Comment by javiercn** [comment]
@AlexanderKot @asamoylik are you both talking about the same app? 

If not, please use separate issues, since it makes it harder to follow up with different threads. 

It'll take some time before we get back to this, since many folks on the team are going to be out of the office for Christmas.

I would suggest applying a constraint to `module` that avoids matching any area value as a first step. The constraint needs to be an inline route constraint. Check for example:
https://github.com/dotnet/aspnetcore/blob/main/src/Http/Routing/src/Constraints/IntRouteConstraint.cs

So the route needs to have something like `{module:modconstraint}` on it for it to do something.

If you want to dig deeper, there is an interface that our constraints implement IParameterLiteralNodeMatchingPolicy that is what we use for reducing the size of the tree in these situations.

On a pattern like
```
context.MapAreaControllerRoute("area6_Virtual", "area6", "{{module}}/area6/{{controller}}/{{action}}/{{id?}}", ...);
context.MapAreaControllerRoute("area6_Default", "area6", "area6/{controller}/{action}/{id?}", ...);
```

If you know that some path like `area6/area6` will never match the `{{module}}/area6/` route, you can use IParameterLiteralNodeMatchingPolicy to avoid creating additional states. Without going into much detail, when our router sees the routes above, it "creates" an "artificial route" "area6/area6/{{controller}}/{{action}}/{{id?}}" and it does that for every literal on a route at that level. (I understand this is not the easiest thing to grasp).

If you know that `{module}` is never going to have the same value as a literal on the same level for any other route, you can implement `IParameterLiteralNodeMatchingPolicy` and return false based on the parameter name and that way "avoid" creating these "artificial routes".

Our constraints do something similar but evaluate the constraint against the value from the literal to "trim the route" when the value doesn't pass the constraint (which happens to be the vast majority of the cases)

To summarize:
* Start by applying the constraint as an inline route constraint (otherwise we won't pick it up).
* If you have a custom constraint instead of a built-in constraint, you need to implement `IParameterLiteralNodeMatchingPolicy`
  * For guidance on the implementation, check how its done in any of the existing built-in constraints.
* If you want to have more control, you can implement your own `IParameterLiteralNodeMatchingPolicy` and use the parameter name to make decissions. (For example, to say that a parameter will never match a literal value on your routes).

**Comment by asamoylik** [comment]
Hello @javiercn, thank you for the suggestions.
Yes, we are talking about the same app.
After moving to inline constraints and implementing `IParameterLiteralNodeMatchingPolicy` memory consumption felled down to ~2.2G. When doubling routes with `{tenant}` parameter at first place memory consumption is ~3.2G. So it's much better than it was on .NET 5.0.12
It would be very helpful if these tips were in the documentation.

> If you want to have more control, you can implement your own `IParameterLiteralNodeMatchingPolicy` and use the parameter name to make decissions. (For example, to say that a parameter will never match a literal value on your routes).

Are you meaning implementing one policy for all parameters instead of constraint per each parameter?

**Comment by AlexanderKot** [comment]
Hello all
I confirm (just for pushing @msftbot to react :) - same project.

**Comment by javiercn** [comment]
@asamoylik that's inline with what I would expect to see.

As an experiment I would suggest you do the following, which is to implement an inline constraint that always returns false and apply that to all the parameters in all routes you have and then measure the memory. 

That would give you an understanding of the minimum amount of memory that you are going to consume. And would mean that you never have a parameter that matches a literal in your application, so you would never match a route that is for example `MyController/MyController` or `MyAction/MyAction`. I think this case is very common, and will work well in most apps, however unfortunately this is not a general assumption we can make.

It's hard for me to provide more guidance without seeing the exact routing configuration, but I would suspect that as long as you can make general assumptions like the ones I made above, you can trim the route table even further. That said, it all depends on the total number of action descriptors the app ends up generating, in cases like this, it might be worth looking at `MapDynamicControllerRoute` since that doesn't create separate endpoints but comes at the cost of having to run action selection on each request (which is much slower).

You can also potentially do a mix and match, where some routes are defined conventionally (which we expand into the router tree and get the "optimal performance" and some other routes you can create as dynamic routes, where they would be slower but use less memory.

**Comment by asamoylik** [comment]
I've done the experiment and added constraint which returns false to all parameters in the route pattern. Our application has 55000 of actions. Minimum amount of memory was around 1.2G. But why we need to add constraints even to controller and action parameters? Application model already has list of available action descriptors.
